### PR TITLE
Langchain autologging support

### PR DESCRIPTION
**Pull request by serena-ruan**
### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?
Langchain flavor autologging support
Depends on https://github.com/langchain-ai/langchain/pull/15691 

### How is this PR tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/langchain/_langchain_autolog.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+            f"{MIN_REQ_VERSION} and {MAX_REQ_VERSION} and may not succeed with packages "
+            "outside this range."
+        )
+
+    config = kwargs.get("config", None)
+    run_id = self.run_id if hasattr(self, "run_id") else None
+    if active_run := mlflow.active_run():
+        run_id = active_run.info.run_id
+        # mlflow callback internally starts the run, so we end it here
+        mlflow.end_run()
```
Comment (BenWilson2) [inline mlflow/langchain/_langchain_autolog.py:83]: Can we validate the functionality for child runs here? Does the callback work with a child run? Will a parent run still be active and what does that do for the langchain callback if this implementation is active?

File: mlflow/ml-package-versions.yml [hunk 2]
```diff
@@ -33,7 +33,7 @@ pytorch:
 
   models:
     minimum: "1.9.0"
-    maximum: "2.1.1"
+    maximum: "2.1.2"
```
Comment (BenWilson2) [inline mlflow/ml-package-versions.yml:36]: Running the update here was just for testing purposes, right? 
For the final PR, we should just add the autologging support version ranges for the current released supported version for langchain and let our CD pipeline handle updating the package version files :)

File: tests/langchain/test_langchain_autolog.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+from mlflow.models.utils import _read_example
+from mlflow.utils.openai_utils import (
+    TEST_CONTENT,
+    _mock_chat_completion_response,
+    _mock_request,
+)
+
+MODEL_DIR = "model"
+TEST_CONTENT = "test"
+MLFLOW_CALLBACK_METRICS = [
```
Comment (BenWilson2) [inline tests/langchain/test_langchain_autolog.py:22]: Are these defined anywhere in langchain so that cross version tests can just read from the list of provided constants from within their library (to simplify our tests if they end up changing these and to make sure that we don't miss new ones in the future)

File: mlflow/langchain/__init__.py [hunk 4]
```diff
@@ -443,12 +453,22 @@ def _load_model(local_model_path, flavor_conf):
...
+    else:
+        raise mlflow.MlflowException(
+            "Failed to load LangChain model. Unknown model type: "
+            f"{flavor_conf.get(_MODEL_TYPE_KEY)}"
+        )
+    # To avoid double logging, we set model_logged to True
+    # when the model is loaded
+    if not autologging_is_disabled(FLAVOR_NAME):
+        if hasattr(model, "__config__"):
+            model.__config__.extra = Extra.allow
```
Comment (BenWilson2) [inline mlflow/langchain/__init__.py:468]: clever :)

File: mlflow/ml-package-versions.yml (hunk 2 above)
Comment (serena-ruan) [reply mlflow/ml-package-versions.yml:36]: Yes will revert this change when merge

File: mlflow/langchain/_langchain_autolog.py (hunk 1 above)
Comment (serena-ruan) [reply mlflow/langchain/_langchain_autolog.py:83]: We could support nested run by checking if there's an active run, then we assume we should create child runs by default. But since we're always just creating a single run, is there any use cases for nested runs? Unless the customer explicitly log another child run, otherwise the parent run will always contain a single child run.

File: tests/langchain/test_langchain_autolog.py (hunk 3 above)
Comment (serena-ruan) [reply tests/langchain/test_langchain_autolog.py:22]: They're not defined inside langchain but we can add it into the file of MLflowCallbackHandler :D

File: mlflow/ml-package-versions.yml (hunk 2 above)
Comment (harupy) [reply mlflow/ml-package-versions.yml:36]: If necessary, you can manually upgrade the langchain versions.

File: mlflow/langchain/_langchain_autolog.py [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+
+def _get_input_data_from_function(func_name, model, args, kwargs):
+    supported_func_names = ["__call__", "invoke"]
+    input_example_exc = None
+    if func_name in supported_func_names:
+        if func_name == "__call__":
+            param_name = "inputs"
+        elif func_name == "invoke":
+            param_name = "input"
+        inference_func = getattr(model, func_name)
```
Comment (BenWilson2) [inline mlflow/langchain/_langchain_autolog.py:35]: What are your thoughts on creating a mapping here?

```python
supported_func_mapping = {
        "__call__": "inputs",
        "invoke": "input"
    }

    # Check if function name is supported
    if func_name in supported_func_mapping:
        param_name = supported_func_mapping[func_name]
        inference_func = getattr(model, func_name)
    ...
```

File: mlflow/langchain/_langchain_autolog.py [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+        self.run_id = mlflow_callback.mlflg.run.info.run_id
+
+    log_inference_history = get_autologging_config(
+        mlflow.langchain.FLAVOR_NAME, "log_inference_history", False
+    )
+    if log_inference_history:
+        if input_example is None:
+            input_data = deepcopy(_get_input_data_from_function(func_name, self, args, kwargs))
+            if input_data is None:
+                _logger.warning("Input data gathering failed, only log inference results.")
```
Comment (BenWilson2) [inline mlflow/langchain/_langchain_autolog.py:226]: should this be an info level in the logger? 

Perhaps we could have the log output be "Failed to log input data, logging only inference results"

I'm curious - is there a chain, agent, or retriever that won't have any input data?

File: tests/langchain/test_langchain_autolog.py [hunk 7]
```diff
@@ -0,0 +1,484 @@
...
+
+        # inference history is appended to the same table
+        with mlflow.start_run(run.info.run_id):
+            model.invoke(question)
+        model.invoke(question)
+        loaded_table = mlflow.load_table("inference_history.json", run_ids=[run.info.run_id])
+        loaded_dict = loaded_table.to_dict("records")
+        assert loaded_dict == [{"input": question, "output": answer}] * 3
+
+        with pytest.raises(MlflowException, match="Please end current run when autologging is on "):
```
Comment (BenWilson2) [inline tests/langchain/test_langchain_autolog.py:325]: Can we not use the active run? Is this because langchain will create a new child run in this instance of having an active run context? 
It would be great if we could allow users to specify their own run here wherein we use their activated run context for logging and if langchain is starting a new run on behalf of users, we disable that functionality iff a run is already active. The user-facing run context is likely where users will be doing things like setting tags, extra requirements, code paths, etc... we most certainly do not want to prevent them from doing that.

File: mlflow/langchain/_langchain_autolog.py [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+        else:
+            input_data = input_example
+        data_dict = _combine_input_and_output(input_data, result)
+        with _wrap_func_with_run(mlflow_callback.mlflg.run.info.run_id):
+            mlflow.log_table(data_dict, "inference_history.json")
```
Comment (BenWilson2) [inline mlflow/langchain/_langchain_autolog.py:231]: If non-jsonnable data is somehow generated, is the exception that would be raised from `log_table()` specific enough to guide users to what went wrong? Can we verify this?

File: mlflow/langchain/_langchain_autolog.py [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+        return {"input": [input], "output": [output]}
+    if isinstance(input, list) and (
+        all(isinstance(x, str) for x in input) or all(isinstance(x, dict) for x in input)
+    ):
+        if not isinstance(output, list) or len(output) != len(input):
+            raise MlflowException(
+                "Failed to combine input and output data with different lengths "
+                "into a single pandas DataFrame. "
+            )
+        return {"input": input, "output": output}
```
Comment (BenWilson2) [inline mlflow/langchain/_langchain_autolog.py:72]: Did we want to create a session_id here to delineate separate instantiations of this model that have been loaded in another session / environment?

File: tests/langchain/test_langchain_autolog.py (hunk 7 above)
Comment (BenWilson2) [reply tests/langchain/test_langchain_autolog.py:325]: Let's use the client API :)

File: mlflow/langchain/_langchain_autolog.py (hunk 6 above)
Comment (serena-ruan) [reply mlflow/langchain/_langchain_autolog.py:219]: BaseRetriever uses `get_relevant_documents` function which accepts `query`, ChatPromptTemplate uses `format_messages` function etc.. But for `Chain` and `AgentExecutor`(also Chain) they should either uses `__call__` or `invoke` function are they're patched.

File: mlflow/langchain/_langchain_autolog.py (hunk 8 above)
Comment (serena-ruan) [reply mlflow/langchain/_langchain_autolog.py]: `log_table` checks below:
```
if not isinstance(data, (pd.DataFrame, dict)):
            raise MlflowException.invalid_parameter_value(
                "data must be a pandas.DataFrame or a dictionary"
            )
```
So it should be fine

**Comment by harupy** [comment]
I manually ran the following code:


```python
from langchain.chains import LLMChain
from langchain.llms import OpenAI
from langchain.prompts import PromptTemplate

import mlflow

mlflow.langchain.autolog()

llm = OpenAI(temperature=0.9)
prompt = PromptTemplate(
    input_variables=["product"],
    template="What is a good name for a company that makes {product}?",
)
chain = LLMChain(llm=llm, prompt=prompt)

print(chain.invoke("soup"))
```

and got this error:

```
2024/01/16 18:26:54 WARNING mlflow.utils.autologging_utils: Encountered unexpected error during autologging: __init__() got an unexpected keyword argument 'run_id'
Error in _MLflowLangchainCallback.on_chain_start callback: AttributeError("'_MLflowLangchainCallback' object has no attribute 'ignore_chain_'")
Error in _MLflowLangchainCallback.on_text callback: AttributeError("'_MLflowLangchainCallback' object has no attribute 'metrics'")
Error in _MLflowLangchainCallback.on_llm_start callback: AttributeError("'_MLflowLangchainCallback' object has no attribute 'ignore_llm_'")
Error in _MLflowLangchainCallback.on_llm_end callback: AttributeError("'_MLflowLangchainCallback' object has no attribute 'ignore_llm_'")
Error in _MLflowLangchainCallback.on_chain_end callback: AttributeError("'_MLflowLangchainCallback' object has no attribute 'ignore_chain_'")
2024/01/16 18:26:55 WARNING mlflow.utils.autologging_utils: Encountered unexpected error during langchain autologging: '_MLflowLangchainCallback' object has no attribute 'mlflg'
{'product': 'soup', 'text': '\n\n1. "Savor Soups Co."\n2. "Bowl of Comfort"\n3. "Soup\'s On Inc."\n4. "The Soup Kitchen Co."\n5. "Hearty Harvest Soups"\n6. "Soulful Soups Co."\n7. "Spoonful of Happiness"\n8. "Homestyle Soup Co."\n9. "Broth Brothers"\n10. "Puree Perfection Soups"'}
```

**Comment by serena-ruan** [comment]
> I manually ran the following code:
> 
> ```python
> from langchain.chains import LLMChain
> from langchain.llms import OpenAI
> from langchain.prompts import PromptTemplate
> 
> import mlflow
> 
> mlflow.langchain.autolog()
> 
> llm = OpenAI(temperature=0.9)
> prompt = PromptTemplate(
>     input_variables=["product"],
>     template="What is a good name for a company that makes {product}?",
> )
> chain = LLMChain(llm=llm, prompt=prompt)
> 
> print(chain.invoke("soup"))
> ```
> 
> and got this error:
> 
> ```
> 2024/01/16 18:26:54 WARNING mlflow.utils.autologging_utils: Encountered unexpected error during autologging: __init__() got an unexpected keyword argument 'run_id'
> Error in _MLflowLangchainCallback.on_chain_start callback: AttributeError("'_MLflowLangchainCallback' object has no attribute 'ignore_chain_'")
> Error in _MLflowLangchainCallback.on_text callback: AttributeError("'_MLflowLangchainCallback' object has no attribute 'metrics'")
> Error in _MLflowLangchainCallback.on_llm_start callback: AttributeError("'_MLflowLangchainCallback' object has no attribute 'ignore_llm_'")
> Error in _MLflowLangchainCallback.on_llm_end callback: AttributeError("'_MLflowLangchainCallback' object has no attribute 'ignore_llm_'")
> Error in _MLflowLangchainCallback.on_chain_end callback: AttributeError("'_MLflowLangchainCallback' object has no attribute 'ignore_chain_'")
> 2024/01/16 18:26:55 WARNING mlflow.utils.autologging_utils: Encountered unexpected error during langchain autologging: '_MLflowLangchainCallback' object has no attribute 'mlflg'
> {'product': 'soup', 'text': '\n\n1. "Savor Soups Co."\n2. "Bowl of Comfort"\n3. "Soup\'s On Inc."\n4. "The Soup Kitchen Co."\n5. "Hearty Harvest Soups"\n6. "Soulful Soups Co."\n7. "Spoonful of Happiness"\n8. "Homestyle Soup Co."\n9. "Broth Brothers"\n10. "Puree Perfection Soups"'}
> ```

Because the change depends on langchain mlflowcallback change, we need to manually install langchain_community wheel like my change in cross-version-tests.yml
```
pip install poetry
git clone --recursive https://github.com/serena-ruan/langchain --depth=2 --branch mlflow_callback /tmp/langchain
cd /tmp/langchain/libs/community
poetry build
pip install dist/langchain_community-0.0.12-py3-none-any.whl --force-reinstall --no-deps (no need to add those two tags if you don't install it already)
```

**Comment by harupy** [comment]
I actually installed langchain-community from your branch. Let me try one more time.

**Comment by serena-ruan** [comment]
> I actually installed langchain-community from your branch. Let me try one more time.

I tested in this notebook: https://e2-dogfood.staging.cloud.databricks.com/?o=6051921418418893#notebook/187185339621970

File: mlflow/langchain/_langchain_autolog.py [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+        "get_relevant_documents": "query",
+    }
+    input_example_exc = None
+    if func_name in func_param_name_mapping:
+        param_name = func_param_name_mapping[func_name]
+        inference_func = getattr(model, func_name)
+        try:
+            # A guard to make sure `param_name` is the first argument of inference function
+            assert next(iter(inspect.signature(inference_func).parameters.keys())) == param_name
+            return kwargs[param_name] if param_name in kwargs else args[0]
```
Comment (harupy) [inline mlflow/langchain/_langchain_autolog.py:45]:
```suggestion
            return kwargs.get(param_name, args[0])
```

File: mlflow/langchain/_langchain_autolog.py [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+        except Exception as e:
+            input_example_exc = e
+    else:
+        input_example_exc = MlflowException(
+            f"Unsupported inference function. Only support {list(func_param_name_mapping.keys())}."
+        )
+    _logger.warning(
+        f"Failed to gather input example of model {model.__class__.__name__}"
+        + f" due to {input_example_exc}."
```
Comment (harupy) [inline mlflow/langchain/_langchain_autolog.py:54]:
```suggestion
        f"Failed to gather input example of model {model.__class__.__name__}"
        f" due to {input_example_exc}."
```

File: mlflow/ml_package_versions.py [hunk 12]
```diff
@@ -21,11 +21,11 @@
         },
         "models": {
             "minimum": "1.9.0",
-            "maximum": "2.1.1"
+            "maximum": "2.1.2"
```
Comment (harupy) [inline mlflow/ml_package_versions.py:24]: can we revert unrelated changes?

File: mlflow/tracking/fluent.py [hunk 13]
```diff
@@ -1334,8 +1350,9 @@ def load_table(
     return MlflowClient().load_table(experiment_id, artifact_file, run_ids, extra_columns)
 
 
-def _record_logged_model(mlflow_model):
-    run_id = _get_or_start_run().info.run_id
+def _record_logged_model(mlflow_model, run_id=None):
+    if run_id is None:
+        run_id = _get_or_start_run().info.run_id
```
Comment (harupy) [inline mlflow/tracking/fluent.py:1355]:
```suggestion
    run_id = run_id or _get_or_start_run().info.run_id
```

File: mlflow/langchain/__init__.py [hunk 14]
```diff
@@ -654,3 +679,84 @@ def load_model(model_uri, dst_path=None):
...
+    :param extra_tags: A dictionary of extra tags to set on each managed run created by autologging.
+    """
+
+    with contextlib.suppress(ImportError):
+        from langchain.agents.agent import AgentExecutor
+        from langchain.chains.base import Chain
+        from langchain.schema import BaseRetriever
+
+        classes = lc_runnables_types() + (AgentExecutor, Chain)
+        for clazz in classes:
```
Comment (daniellok-db) [inline mlflow/langchain/__init__.py:747]: maybe `cls` is more idiomatic here?

File: mlflow/langchain/_langchain_autolog.py [hunk 15]
```diff
@@ -0,0 +1,296 @@
...
+        if input_example_exc
+        else ""
```
Comment (daniellok-db) [inline mlflow/langchain/_langchain_autolog.py:56]: is there a case where we reach here but `input_example_exc` is not defined?

File: mlflow/langchain/_langchain_autolog.py [hunk 16]
```diff
@@ -0,0 +1,296 @@
...
+    except ImportError as e:
+        warnings.warn(
+            "MLflow langchain autologging might log model several "
+            "times due to the pydantic.config.Extra import error. "
+            f"Error: {e}"
+        )
+        return False
+    else:
+        if hasattr(model, "__config__"):
+            model.__config__.extra = Extra.allow
```
Comment (daniellok-db) [inline mlflow/langchain/_langchain_autolog.py:97]: should we add some comment to explain why it's necessary to override the pydantic config here?

File: mlflow/langchain/_langchain_autolog.py [hunk 17]
```diff
@@ -0,0 +1,296 @@
...
+    - data
+
+    We patch either `invoke` or `__call__` function for different models
+    based on their usage.
+    """
+
+    import langchain
+
+    # import from langchain_community for test purpose
+    from langchain_community.callbacks import MlflowCallbackHandler
```
Comment (harupy) [inline mlflow/langchain/_langchain_autolog.py:190]: `MlflowCallbackHandler` logs parameters, metrics, and artifacts?

File: mlflow/langchain/_langchain_autolog.py [hunk 18]
```diff
@@ -0,0 +1,296 @@
...
+
+    _lc_version = Version(langchain.__version__)
+    if not MIN_REQ_VERSION <= _lc_version <= MAX_REQ_VERSION:
+        warnings.warn(
+            "Autologging is known to be compatible with langchain versions between "
+            f"{MIN_REQ_VERSION} and {MAX_REQ_VERSION} and may not succeed with packages "
+            "outside this range."
+        )
+
+    run_id = self.run_id if hasattr(self, "run_id") else None
```
Comment (harupy) [inline mlflow/langchain/_langchain_autolog.py:210]: /ai rewrite this using getattr

File: mlflow/langchain/_langchain_autolog.py [hunk 19]
```diff
@@ -0,0 +1,296 @@
...
+
+            registered_model_name = get_autologging_config(
+                mlflow.langchain.FLAVOR_NAME, "registered_model_name", None
+            )
+            extra_tags = get_autologging_config(mlflow.langchain.FLAVOR_NAME, "extra_tags", None)
+            tags = _resolve_extra_tags(mlflow.langchain.FLAVOR_NAME, extra_tags)
+            # self manage the run as we need to get the run_id from mlflow_callback
+            # only log the tags once the first time we log the model
+            for key, value in tags.items():
+                mlflow.MlflowClient().set_tag(mlflow_callback.mlflg.run_id, key, value)
```
Comment (harupy) [inline mlflow/langchain/_langchain_autolog.py:261]: can we use `set_tags`?

File: mlflow/langchain/_langchain_autolog.py [hunk 20]
```diff
@@ -0,0 +1,296 @@
...
+
+def _get_input_data_from_function(func_name, model, args, kwargs):
+    func_param_name_mapping = {
+        "__call__": "inputs",
+        "invoke": "input",
+        "get_relevant_documents": "query",
+    }
+    input_example_exc = None
+    if func_name in func_param_name_mapping:
+        param_name = func_param_name_mapping[func_name]
```
Comment (harupy) [inline mlflow/langchain/_langchain_autolog.py:40]:
```suggestion
    if param_name := func_param_name_mapping.get(func_name):
```

File: mlflow/langchain/_langchain_autolog.py [hunk 21]
```diff
@@ -0,0 +1,296 @@
...
+        inference_func = getattr(model, func_name)
+        try:
```
Comment (harupy) [inline mlflow/langchain/_langchain_autolog.py:42]: Do we need try-catch here? Can we `if`?

```
if next(iter(inspect.signature(inference_func).parameters.keys())) == param_name:
    ...
```

File: mlflow/langchain/_langchain_autolog.py [hunk 22]
```diff
@@ -0,0 +1,296 @@
...
+            model.__config__.extra = Extra.allow
+        return True
+
+
+def _inject_mlflow_callback(func_name, mlflow_callback, args, kwargs):
+    if func_name == "invoke":
+        from langchain.schema.runnable.config import RunnableConfig
+
+        in_args = False
+        # `config` is the second positional argument of runnable.invoke function
```
Comment (harupy) [inline mlflow/langchain/_langchain_autolog.py:106]: A link to the API reference of `invoke` might be helpful.

File: mlflow/langchain/_langchain_autolog.py [hunk 23]
```diff
@@ -0,0 +1,296 @@
...
+    run_id = self.run_id if hasattr(self, "run_id") else None
+    if (active_run := mlflow.active_run()) and run_id is None:
+        run_id = active_run.info.run_id
+    # TODO: test adding callbacks works
+    mlflow_callback = _MLflowLangchainCallback(
+        tracking_uri=mlflow.get_tracking_uri(),
+        run_id=run_id,
+    )
+    args, kwargs = _inject_mlflow_callback(func_name, mlflow_callback, args, kwargs)
+    with disable_autologging():
```
Comment (daniellok-db) [inline mlflow/langchain/_langchain_autolog.py:219]: do we disable autologging when running the inference function because the logging is handled by the `mlflow_callback` injection?

File: mlflow/langchain/_langchain_autolog.py [hunk 24]
```diff
@@ -0,0 +1,296 @@
...
+        else:
+            kwargs["config"] = config
+        return args, kwargs
+
+    if func_name == "__call__":
+        # `callbacks` is the third positional argument of chain.__call__ function
+        if len(args) >= 3:
+            callbacks = args[2] or []
+            callbacks.append(mlflow_callback)
+            args = args[:2] + (callbacks,) + args[3:]
```
Comment (harupy) [inline mlflow/langchain/_langchain_autolog.py:130]:
```suggestion
        if len(args) >= 3:
            (first, second, callbacks, *rest) = args
            args = (first, second, (callbacks or []) + [mlflow_callback], *rest)
```

File: mlflow/langchain/_langchain_autolog.py [hunk 25]
```diff
@@ -0,0 +1,296 @@
...
+        else:
+            callbacks = kwargs.get("callbacks") or []
```
Comment (harupy) [inline mlflow/langchain/_langchain_autolog.py:132]:
```suggestion
            callbacks = kwargs.get("callbacks", [])
```

File: mlflow/langchain/_langchain_autolog.py (hunk 24 above)
Comment (harupy) [reply mlflow/langchain/_langchain_autolog.py:130]: Never mind. I think the original code is more readable.

File: mlflow/langchain/_langchain_autolog.py (hunk 19 above)
Comment (harupy) [reply mlflow/langchain/_langchain_autolog.py:261]: I see, `MlflowClient` doesn't have `set_tags`

File: mlflow/ml-package-versions.yml [hunk 26]
```diff
@@ -577,6 +577,27 @@ langchain:
...
+          "google-search-results",
+          "psutil",
+          "faiss-cpu",
+          "langchain-experimental",
+          "numexpr",
+          "spacy",
+        ]
+    run: |
+      # For MlflowCallback text analysis
+      python -m spacy download en_core_web_sm
```
Comment (harupy) [inline mlflow/ml-package-versions.yml:599]: where is dataset used?

File: mlflow/langchain/_langchain_autolog.py (hunk 17 above)
Comment (serena-ruan) [reply mlflow/langchain/_langchain_autolog.py:190]: It logs metrics, no parameters as all the parameters are part of the logged model

File: mlflow/langchain/_langchain_autolog.py (hunk 25 above)
Comment (serena-ruan) [reply mlflow/langchain/_langchain_autolog.py:132]: This actually prevents situation where callbacks=None

File: mlflow/langchain/_langchain_autolog.py (hunk 23 above)
Comment (serena-ruan) [reply mlflow/langchain/_langchain_autolog.py:219]: It's to avoid patch internal functions that's called within the inference function, for runnable sequence as an example, it invokes each step's inference function, if we don't disable autologging then each step would log metrics/models and it messes up

File: mlflow/ml-package-versions.yml (hunk 26 above)
Comment (serena-ruan) [reply mlflow/ml-package-versions.yml:599]: It's used inside MLflowCallbackHandler's analyze_text function https://github.com/langchain-ai/langchain/blob/7d444724d7582386de347fb928619c2243bd0e55/libs/community/langchain_community/callbacks/mlflow_callback.py#L76

Comment (serena-ruan) [reply mlflow/ml-package-versions.yml:599]: https://github.com/langchain-ai/langchain/blob/7d444724d7582386de347fb928619c2243bd0e55/libs/community/langchain_community/callbacks/mlflow_callback.py#L272

Comment (serena-ruan) [reply mlflow/ml-package-versions.yml:599]: Actually I think I should add a warning at that line to ask user to run this command if `spacy.load("en_core_web_sm")` fails, and wrap it into a try catch

File: mlflow/langchain/__init__.py (hunk 14 above)
Comment (serena-ruan) [reply mlflow/langchain/__init__.py:747]: Other functions use clazz as well lol but I'll adapt your advice :D

File: .github/workflows/cross-version-tests.yml [hunk 27]
```diff
@@ -188,6 +188,15 @@ jobs:
...
         run: |
           ${{ matrix.install }}
+          # temp workaround to install from feature branch
+          if [ "${{ matrix.package }}/${{ matrix.category }}" = "langchain/autologging" ]
+          then
+            pip install poetry
+            git clone --recursive https://github.com/serena-ruan/langchain --depth=2 --branch mlflow_callback /tmp/langchain
+            cd /tmp/langchain/libs/community
+            poetry build
+            pip install $(ls dist/*.whl) --force-reinstall --no-deps
```
[the remark this answers is not in the document]
Comment (harupy) [reply .github/workflows/cross-version-tests.yml:198]: sounds good

File: mlflow/langchain/_langchain_autolog.py (hunk 25 above)
Comment (harupy) [reply mlflow/langchain/_langchain_autolog.py:136]: ah makes sense!

File: mlflow/langchain/_langchain_autolog.py (hunk 17 above)
Comment (harupy) [reply mlflow/langchain/_langchain_autolog.py:195]: Got it, thanks for the explanation!

File: mlflow/ml-package-versions.yml (hunk 26 above)
Comment (harupy) [reply mlflow/ml-package-versions.yml:599]: I see, so MLflowCallbackHandler depends on the `en_core_web_sm` dataset existing?

Comment (serena-ruan) [reply mlflow/ml-package-versions.yml:599]: Yes

**Review by harupy — APPROVED** [review]
LGTM!

**Comment by serena-ruan** [comment]
@BenWilson2 Can I merge this for 2.10.0? Or should we wait for next release?

**Review by BenWilson2 — APPROVED** [review]
LGTM! Let's include in 2.10! Great work :)
